File: go/src/deposit/src/structs.go
package deposit

import (
	"errors"
	"interview_mock_deposits_go/deposit/src/domain"
	"interview_mock_deposits_go/internal/testUtils"
	"interview_mock_deposits_go/utils"
)

type Transaction struct {
	UserId        string                 `json:"userId"`
	Type          domain.TransactionType `json:"type"`
	TransactionId string                 `json:"transactionId"`
	Amount        float64                `json:"amount"`
	DelayHours    *int                   `json:"delayHours,omitempty"`
}

type Deposit struct {
	DepositId string               `json:"depositId"`
	Status    domain.DepositStatus `json:"status"`
	Transaction
	UserData    *UserData       `json:"userData,omitempty"`
	DepositTime *utils.DateTime `json:"depositTime,omitempty"`
	CreatedAt   string          `json:"createdAt,omitempty"`
}

type UserData struct {
	Name        string      `json:"name"`
	Document    string      `json:"document"`
	BankAccount BankAccount `json:"bankAccount"`
}

func (t UserData) Validate() error {
	if t.Document == "" {
		return errors.New("invalid document")
	}
	if t.Name == "" {
		return errors.New("invalid name")
	}
	return t.BankAccount.Validate()
}

type BankAccount struct {
	Bank    string `json:"bank"`
	Branch  string `json:"branch"`
	Account string `json:"account"`
}

func (t BankAccount) Validate() error {
	if t.Bank == "" {
		return errors.New("invalid bank")
	}
	if t.Branch == "" {
		return errors.New("invalid bank branch")
	}
	if t.Account == "" {
		return errors.New("invalid bank account")
	}
	return nil
}

func (t Transaction) GenerateIdempotencyKey() string {
	return testUtils.GenerateRandomId()
}

func (t Transaction) GenerateDepositId() string {
	return testUtils.GenerateRandomId()
}

File: python/src/deposit/structs.py
import uuid
from dataclasses import dataclass, field

from deposit.domain.deposit_status import DepositStatus
from deposit.domain.transaction_type import TransactionType


@dataclass
class BankAccount:
    bank: str
    branch: str
    account: str

    def validate(self) -> None:
        if not self.bank:
            raise ValueError("invalid bank")
        if not self.branch:
            raise ValueError("invalid bank branch")
        if not self.account:
            raise ValueError("invalid bank account")


@dataclass
class UserData:
    name: str
    document: str
    bank_account: BankAccount

    def validate(self) -> None:
        if not self.document:
            raise ValueError("invalid document")
        if not self.name:
            raise ValueError("invalid name")
        self.bank_account.validate()


@dataclass
class Transaction:
    user_id: str
    type: TransactionType
    transaction_id: str
    amount: float
    delay_hours: int | None = None

    def generate_deposit_id(self) -> str:
        return str(uuid.uuid4())


@dataclass
class Deposit:
    deposit_id: str
    status: DepositStatus
    transaction: Transaction
    user_data: UserData | None = None
    created_at: str | None = field(default=None)
